File: scripts/inventory-workbook.py
"""Convert cached Balance values to a reviewable opening-stock CSV. Never writes a database.
Usage: python scripts/inventory-workbook.py source.xlsx --out artifacts/inventory-import
"""
import argparse, csv, datetime, json, re, zipfile
from collections import Counter
from decimal import Decimal, ROUND_HALF_UP
from pathlib import Path
import xml.etree.ElementTree as ET

NS = {'s': 'http://schemas.openxmlformats.org/spreadsheetml/2006/main'}

def workbook(path):
    with zipfile.ZipFile(path) as archive:
        strings = [''.join(t.text or '' for t in si.iter('{'+NS['s']+'}t')) for si in ET.fromstring(archive.read('xl/sharedStrings.xml'))] if 'xl/sharedStrings.xml' in archive.namelist() else []
        rels = {r.attrib['Id']: r.attrib['Target'] for r in ET.fromstring(archive.read('xl/_rels/workbook.xml.rels'))}
        result = {}
        for sheet in ET.fromstring(archive.read('xl/workbook.xml')).find('s:sheets', NS):
            target = rels[sheet.attrib['{http://schemas.openxmlformats.org/officeDocument/2006/relationships}id']]
            root = ET.fromstring(archive.read(target.lstrip('/') if target.startswith('/') else 'xl/'+target))
            rows = []
            for row in root.findall('s:sheetData/s:row', NS):
                cells = {}
                for cell in row:
                    value = cell.find('s:v', NS)
                    value = value.text if value is not None else ''
                    if cell.attrib.get('t') == 's': value = strings[int(value)] if value else ''
                    elif cell.attrib.get('t') == 'inlineStr': value = ''.join(t.text or '' for t in cell.findall('.//s:t', NS))
                    if value: cells[re.sub(r'\d', '', cell.attrib['r'])] = value
                if cells: rows.append((int(row.attrib['r']), cells))
            result[sheet.attrib['name']] = rows
        return result

def money(value):
    return str(Decimal(value or '0').quantize(Decimal('.01'), rounding=ROUND_HALF_UP))

def identifier(value):
    if re.fullmatch(r'\d+(\.\d+)?[Ee][+\-]?\d+|\d+\.0+', value or ''):
        return format(Decimal(value), '.0f')
    return value.strip()

def convert(source, destination):
    sheets = workbook(source)
    if 'Balance' not in sheets: raise ValueError('Balance sheet олдсонгүй.')
    fields = ['code','brand','supplier','name','capacity','color','imei','warehouse','qty','unit_cost','total_cost','sale_price','variant','min_stock']
    output, issues, seen = [], [], set()
    for number, r in sheets['Balance']:
        if number <= 2: continue
        if not r.get('B') and not r.get('D'): continue  # formatting / formula-only trailing rows
        if not all(r.get(k) for k in ['B','D','H']):
            issues.append({'row':number,'reason':'Код, нэр эсвэл агуулах дутуу','code':r.get('B','')}); continue
        try:
            qty = Decimal(r.get('P','0'))
            if qty < 0 or qty != int(qty): raise ValueError('Үлдэгдэл сөрөг эсвэл бүхэл бус')
            code = identifier(r['B'])
            if code in seen: raise ValueError('Код давхардсан')
            total = money(r.get('R','0'))
            if any(Decimal(money(r.get(k,'0'))) < 0 for k in ['Q','R','S']): raise ValueError('Өртөг / үнэ сөрөг')
            if qty == 0 and Decimal(total) != 0: raise ValueError('Тэг үлдэгдэлтэй боловч өртөгтэй')
            output.append(dict(zip(fields,[code,r.get('C','').strip(),r.get('C','').strip(),r['D'].strip(),r.get('E',''),r.get('F',''),identifier(r.get('G','')),r['H'].strip(),int(qty),money(r.get('Q','0')),total,money(r.get('S','0')),'',0])))
            seen.add(code)
        except (ValueError, ArithmeticError) as error:
            issues.append({'row':number,'reason':str(error),'code':r.get('B','')})
    destination.mkdir(parents=True, exist_ok=True)
    with (destination/'opening-stock.csv').open('w',encoding='utf-8-sig',newline='') as handle:
        writer=csv.DictWriter(handle,fieldnames=fields);writer.writeheader();writer.writerows(output)
    first=sheets['Balance'][0][1]; serial=Decimal(first.get('R',first.get('Q','0')))
    as_of=(datetime.datetime(1899,12,30)+datetime.timedelta(days=float(serial))).isoformat()+'+08:00'
    summary={'source':source.name,'sheets':[{'name':name,'nonempty_rows':len(rows)} for name,rows in sheets.items()], 'as_of_exclusive':as_of,'products':len(output),'positive_products':sum(r['qty']>0 for r in output),'units':sum(r['qty'] for r in output),'value':str(sum(Decimal(r['total_cost']) for r in output)),'warehouses':dict(Counter(r['warehouse'] for r in output)),'issues':issues,'note':'Cached Balance closing values only. Historical Purchase/Sales rows are not replayed. Review issues and as-of date before import.'}
    (destination/'review.json').write_text(json.dumps(summary,ensure_ascii=False,indent=2),encoding='utf-8')
    print(json.dumps({k:v for k,v in summary.items() if k!='sheets'},ensure_ascii=False))
    return summary

if __name__ == '__main__':
    parser=argparse.ArgumentParser();parser.add_argument('source',type=Path);parser.add_argument('--out',type=Path,required=True)
    args=parser.parse_args();convert(args.source,args.out)

File: lib/inventory-workbook.ts
import {openingRows} from './inventory-csv';
export type ImportFile={rows:Record<string,unknown>[];warnings:string[];asOf?:string;sourceRows:number};
const identifier=(s:string)=>/^(\d+(\.\d+)?[Ee][+\-]?\d+|\d+\.0+)$/.test(s)&&Number.isSafeInteger(Number(s))?Number(s).toFixed(0):s.trim();
const amount=(s:string|undefined)=>Math.round(Number(s||0)*100)/100;
export function balanceRows(rows:{number:number;cells:Record<string,string>}[]):ImportFile {
 const result:ImportFile={rows:[],warnings:[],sourceRows:0},seen=new Set<string>();
 const report=rows.find(r=>r.number===1)?.cells;
 if(report?.R){const end=Date.UTC(1899,11,30)+Number(report.R)*86400000-1;result.asOf=new Date(end).toISOString().slice(0,16);}
 for(const {number,cells:r} of rows){
  if(number<=2||(!r.B&&!r.D))continue;result.sourceRows++;
  const code=identifier(r.B||''),qty=Number(r.P||0),total=amount(r.R);
  let issue='';
  if(!code||!r.D?.trim()||!r.H?.trim())issue='Код, нэр эсвэл агуулах дутуу';
  else if(seen.has(code))issue='Зайг цэвэрлэсний дараа код давхардсан';
  else if(r.P===undefined||r.P===''||r.R===undefined||r.R==='')issue='Эцсийн үлдэгдэл / өртгийн тооцоолсон утга дутуу. Excel-д дахин тооцоолж хадгална уу';
  else if(!Number.isSafeInteger(qty)||qty<0)issue='Үлдэгдэл сөрөг эсвэл бүхэл бус';
  else if([total,amount(r.Q),amount(r.S)].some(n=>!Number.isFinite(n)||n<0))issue='Өртөг / үнэ буруу';
  else if(!qty&&total!==0)issue='Тэг үлдэгдэлтэй боловч өртөгтэй';
  if(issue){result.warnings.push(`Balance ${number}-р мөр (${code||'кодгүй'}): ${issue}`);continue;}
  seen.add(code);
  result.rows.push({code,brand:(r.C||'').trim(),supplier:(r.C||'').trim(),name:r.D.trim(),capacity:r.E||'',color:r.F||'',imei:identifier(r.G||''),warehouse:r.H.trim(),qty,unit_cost:amount(r.Q),total_cost:total,sale_price:amount(r.S),variant:'',min_stock:0});
 }
 if(!result.rows.length)throw new Error('Balance sheet-д импортлох бараа олдсонгүй.');
 return result;
}
export async function readInventoryFile(file:File):Promise<ImportFile>{
 if(file.size>10_000_000)throw new Error('10 MB-аас бага файл сонгоно уу.');
 if(file.name.toLowerCase().endsWith('.csv')){const rows=openingRows(await file.text());return {rows,warnings:[],sourceRows:rows.length};}
 if(!file.name.toLowerCase().endsWith('.xlsx'))throw new Error('.xlsx эсвэл .csv файл сонгоно уу.');
 const {unzipSync,strFromU8}=await import('fflate');let total=0;
 const archive=unzipSync(new Uint8Array(await file.arrayBuffer()),{filter:entry=>{
  if(!/^xl\/(workbook\.xml|_rels\/workbook\.xml\.rels|sharedStrings\.xml|worksheets\/sheet\d+\.xml)$/.test(entry.name))return false;
  total+=entry.originalSize;if(total>60_000_000||entry.originalSize>25_000_000)throw new Error('Excel-ийн задлагдсан хэмжээ хэт том.');return true;
 }});
 const xml=(path:string)=>{
  if(!archive[path])throw new Error('Excel-ийн бүтэц дутуу: '+path);
  const text=strFromU8(archive[path]);if(/<!DOCTYPE/i.test(text))throw new Error('Excel XML бүтэц дэмжигдэхгүй.');
  const doc=new DOMParser().parseFromString(text,'application/xml');if(doc.querySelector('parsererror'))throw new Error('Excel XML бүтэц буруу.');return doc;
 };
 const book=xml('xl/workbook.xml'),sheet=[...book.getElementsByTagName('sheet')].find(s=>s.getAttribute('name')==='Balance');
 if(!sheet)throw new Error('Balance нэртэй sheet олдсонгүй.');
 const rid=sheet.getAttribute('r:id'),rel=[...xml('xl/_rels/workbook.xml.rels').getElementsByTagName('Relationship')].find(r=>r.getAttribute('Id')===rid);
 const target=rel?.getAttribute('Target')||'',path=target.startsWith('/')?target.slice(1):'xl/'+target;
 const strings=archive['xl/sharedStrings.xml']?[...xml('xl/sharedStrings.xml').getElementsByTagName('si')].map(si=>[...si.getElementsByTagName('t')].map(t=>t.textContent||'').join('')):[];
 const rows=[...xml(path).getElementsByTagName('row')].map(row=>({number:Number(row.getAttribute('r')),cells:Object.fromEntries([...row.getElementsByTagName('c')].map(cell=>{
  let value=cell.getElementsByTagName('v')[0]?.textContent||'';
  if(cell.getAttribute('t')==='s')value=strings[Number(value)]||'';
  if(cell.getAttribute('t')==='inlineStr')value=cell.textContent||'';
  return [(cell.getAttribute('r')||'').replace(/\d/g,''),value];
 }))}));
 return balanceRows(rows);
}
